File: 160-Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
#  O(lenA + lenB)
        pA, pB = headA, headB
        while pA != pB:
            pA = headB if not pA else pA.next
            pB = headA if not pB else pB.next
        
        return pA

File: JavaScript/160-Intersection of Two Linked Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */

// use set
var getIntersectionNode = function(headA, headB) {
    let visited = new Set();
    while (headA){
        visited.add(headA);
        headA = headA.next;
    }
    
    while (headB){
        if (visited.has(headB)){
            return headB;
        }
        headB = headB.next;
    }
    
    return null;
};


// without extra space
var getIntersectionNode = function(headA, headB) {
    let pA = headA, pB = headB;
    while (pA != pB){
        if (pA === null) pA = headB;
        else pA = pA.next;
        
        if (pB === null) pB = headA;
        else pB = pB.next;
    }
    
    return pA;
};
